pkg/apis/pravega/v1alpha1/pravega.go

```diff
@@ -11,35 +11,175 @@
+	// is present and it will use it as Tier 2
+	Tier2 *Tier2Spec `json:"tier2"`
+
+	// ControllerServiceAccountName configures the service account used on controller instances.
+	// If not specified, Kubernetes will automatically assign the default service account in the namespace
+	ControllerServiceAccountName string `json:"controllerServiceAccountName,omitempty"`
+
+	// SegmentStoreServiceAccountName configures the service account used on segment store instances.
+	// If not specified, Kubernetes will automatically assign the default service account in the namespace
+	SegmentStoreServiceAccountName string `json:"segmentStoreServiceAccountName,omitempty"`
 }
 
-type MetricsSpec struct {
+func (s *PravegaSpec) withDefaults() {
+	if s.ControllerReplicas < 1 {
+		s.ControllerReplicas = 1
+	}
+
+	if s.SegmentStoreReplicas < 1 {
+		s.SegmentStoreReplicas = 1
+	}
+
+	if s.Image == nil {
+		s.Image = &PravegaImageSpec{}
+	}
+	s.Image.withDefaults()
+
+	if s.Options == nil {
+		s.Options = map[string]string{}
+	}
+
+	if s.CacheVolumeClaimTemplate == nil {
+		s.CacheVolumeClaimTemplate = &v1.PersistentVolumeClaimSpec{
+			AccessModes: []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce},
+			Resources: v1.ResourceRequirements{
+				Requests: v1.ResourceList{
+					v1.ResourceStorage: resource.MustParse(DefaultPravegaCacheVolumeSize),
+				},
+			},
+		}
+	}
+
+	if s.Tier2 == nil {
+		s.Tier2 = &Tier2Spec{}
+	}
+	s.Tier2.withDefaults()
+}
+
+// PravegaImageSpec defines the fields needed for a Pravega Docker image
+type PravegaImageSpec struct {
+	ImageSpec
+}
+
+// String formats a container image struct as a Docker compatible repository string
+func (s *PravegaImageSpec) String() string {
+	return fmt.Sprintf("%s:%s", s.Repository, s.Tag)
 }
 
+func (s *PravegaImageSpec) withDefaults() {
+	if s.Repository == "" {
+		s.Repository = DefaultPravegaImageRepository
+	}
+
+	if s.Tag == "" {
+		s.Tag = DefaultPravegaImageTag
+	}
+
+	if s.PullPolicy == "" {
+		s.PullPolicy = DefaultPravegaImagePullPolicy
+	}
+}
+
+// Tier2Spec configures the Tier 2 storage type to use with Pravega.
+// If not specified, Tier 2 will be configured in filesystem mode and will try
+// to use a PersistentVolumeClaim with the name "pravega-tier2"
 type Tier2Spec struct {
-	FileSystem *FileSystemSpec `json:"filesystem"`
-	Ecs        *ECSSpec        `json:"ecs"`
-	Hdfs       *HDFSSpec       `json:"hdfs"`
+	// FileSystem is used to configure a pre-created Persistent Volume Claim
+	// as Tier 2 backend.
+	// It is default Tier 2 mode.
+	FileSystem *FileSystemSpec `json:"filesystem,omitempty"`
+
+	// Ecs is used to configure a Dell EMC ECS system as a Tier 2 backend
+	Ecs *ECSSpec `json:"ecs,omitempty"`
+
+	// Hdfs is used to configure an HDFS system as a Tier 2 backend
+	Hdfs *HDFSSpec `json:"hdfs,omitempty"`
+}
+
+func (s *Tier2Spec) withDefaults() {
+	if s.FileSystem == nil && s.Ecs == nil && s.Hdfs == nil {
+		fs := &FileSystemSpec{
```

Comment:
Naïve question: is it possible that there is no file system PV configured? If so, how does the deployment fail?

Reply:
Yes, it is possible to create a `PravegaCluster` that references a non-existent PVC. The deployment will succeed, but the Segment Store pods will fail to start and will remain in `Pending` state until the PVC is created (`pravega-tier2` by default).

Example:

```
$ kubectl get po
...
pravega-pravega-segmentstore-0                0/1     Pending             0          22s
```
A user can describe the pod to find out what the problem is:

```
$ kubectl describe po pravega-pravega-segmentstore-0
...
Events:
  Type     Reason            Age                    From               Message
  ----     ------            ----                   ----               -------
  Warning  FailedScheduling  3m10s (x4 over 3m13s)  default-scheduler  PersistentVolumeClaim is not bound: "cache-pravega-pravega-segmentstore-0" (repeated 4 times)
  Warning  FailedScheduling  7s (x12 over 3m6s)     default-scheduler  persistentvolumeclaim "pravega-tier2" not found (repeated 4 times)
```

Segment Store pods will resume their creation process as soon as the PVC is created.

```
$ kubectl create -f pvc.yaml
$ kubectl get po
...
pravega-pravega-segmentstore-0                0/1     ContainerCreating   0          4m16s
...
pravega-pravega-segmentstore-0                1/1     Running   0          5m4s
```